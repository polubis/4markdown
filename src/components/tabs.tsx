import React, { useState } from 'react';
import c from 'classnames';
import { Button } from 'design-system/button';
import { BiX } from 'react-icons/bi';
interface TabProps {
  label?: string;
  children: React.ReactNode;
  setRemovedTabs: React.Dispatch<React.SetStateAction<string[]>>;
}

const Tab: React.FC<TabProps> = ({ children }) => {
  return <div>{children}</div>;
};

interface TabsProps {
  children: React.ReactNode;
}

const Tabs: React.FC<TabsProps> = ({ children }) => {
  const [activeTab, setActiveTab] = useState(0);

  return (
    <div>
      <div className="tab-headers flex space-x-2">
        {React.Children.map(children, (child, index) => {
          const labelFromChild = (child as React.ReactElement<TabProps>).props
            .label;

          return (
            <Button
              i={2}
              s={2}
              auto
              type="button"
              className={c(
                activeTab === index ? `!text-white` : `!text-gray-400`,
                `group relative  pr-8`,
              )}
              onClick={() => setActiveTab(index)}
            >
              {labelFromChild}
              <BiX
                onClick={() => {
                  if (labelFromChild) {
                    (
                      child as React.ReactElement<TabProps>
                    ).props.setRemovedTabs((prev) => [...prev, labelFromChild]);
                  }
                }}
                className=" group-hover:block hidden transition-all duration-200 ease hover:text-white text-gray-400 absolute top-1/2 right-1 -translate-y-1/2 "
              />
            </Button>
          );
        })}
      </div>
      <div className="tab-content mt-4">
        {React.Children.toArray(children)[activeTab]}
      </div>
    </div>
  );
};

export { Tabs, Tab };
