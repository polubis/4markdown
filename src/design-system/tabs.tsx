import React from 'react';
import c from 'classnames';

interface TabsItemProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
  active?: boolean;
}

interface TabsProps {
  className?: string;
  children: React.ReactElement | React.ReactElement[];
  fit?: boolean;
  loading?: boolean;
}

const Tabs = ({ className, children, fit, loading }: TabsProps) => {
  const enhancedChildren = React.Children.map(
    children,
    (child: React.ReactElement<TabsItemProps>) => {
      if (!React.isValidElement(child))
        throw Error(`Passed children is not an React element`);

      return React.cloneElement(child, { ...child.props, disabled: loading });
    },
  );

  return (
    <div
      className={c(
        className,
        `relative 
        flex rounded-md [&>*:first-child]:rounded-s-md [&>*:last-child]:rounded-r-md`,
        {
          [`w-fit`]: fit,
        },
      )}
    >
      {enhancedChildren}
      {loading && (
        <div className="absolute z-10 top-0 right-0 left-0 bottom-0 bg-slate-300/40 rounded-r-md rounded-s-md " />
      )}
    </div>
  );
};

const Item = ({ active, ...props }: TabsItemProps) => {
  return (
    <button
      className={c(
        `font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30`,
        { [`bg-green-400`]: active },
        {
          [`enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white`]:
            !active,
        },
      )}
      {...props}
    />
  );
};

Tabs.Item = Item;

export { Tabs };
