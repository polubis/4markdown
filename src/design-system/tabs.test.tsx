import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Tabs } from './tabs';
import { expect } from '@jest/globals';

describe(`Tabs works when:`, () => {
  it(`displays selection and items`, () => {
    const spy = jest.fn();

    const { asFragment } = render(
      <Tabs>
        <Tabs.Item active>First</Tabs.Item>
        <Tabs.Item onClick={spy}>Second</Tabs.Item>
        <Tabs.Item>Last</Tabs.Item>
      </Tabs>,
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText(/Second/));

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it(`makes tabs disabled when loading`, () => {
    const spy = jest.fn();

    const { asFragment, rerender } = render(
      <Tabs loading>
        <Tabs.Item active>First</Tabs.Item>
        <Tabs.Item onClick={spy}>Second</Tabs.Item>
        <Tabs.Item>Last</Tabs.Item>
      </Tabs>,
    );

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText(/Second/));

    expect(spy).toHaveBeenCalledTimes(0);

    rerender(
      <Tabs>
        <Tabs.Item active>First</Tabs.Item>
        <Tabs.Item onClick={spy}>Second</Tabs.Item>
        <Tabs.Item>Last</Tabs.Item>
      </Tabs>,
    );

    fireEvent.click(screen.getByText(/Second/));

    expect(spy).toHaveBeenCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });
});
