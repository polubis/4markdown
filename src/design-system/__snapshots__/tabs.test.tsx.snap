// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs works when: displays selection and items 1`] = `
<DocumentFragment>
  <div
    class="relative 
        flex rounded-md [&>*:first-child]:rounded-s-md [&>*:last-child]:rounded-r-md"
  >
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 bg-green-400"
    >
      First
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
    >
      Second
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
    >
      Last
    </button>
  </div>
</DocumentFragment>
`;

exports[`Tabs works when: makes tabs disabled when loading 1`] = `
<DocumentFragment>
  <div
    class="relative 
        flex rounded-md [&>*:first-child]:rounded-s-md [&>*:last-child]:rounded-r-md"
  >
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 bg-green-400"
      disabled=""
    >
      First
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
      disabled=""
    >
      Second
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
      disabled=""
    >
      Last
    </button>
    <div
      class="absolute z-10 top-0 right-0 left-0 bottom-0 bg-slate-300/40 rounded-r-md rounded-s-md "
    />
  </div>
</DocumentFragment>
`;

exports[`Tabs works when: makes tabs disabled when loading 2`] = `
<DocumentFragment>
  <div
    class="relative 
        flex rounded-md [&>*:first-child]:rounded-s-md [&>*:last-child]:rounded-r-md"
  >
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 bg-green-400"
    >
      First
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
    >
      Second
    </button>
    <button
      class="font-medium flex-1 py-2 px-3 disabled:bg-neutral-300/30 disabled:text-black/30 dark:disabled:bg-gray-900 dark:disabled:text-white/30 enabled:hover:bg-gray-400/70 dark:enabled:hover:bg-slate-800/70 dark:bg-slate-800 bg-gray-300 text-black dark:text-white"
    >
      Last
    </button>
  </div>
</DocumentFragment>
`;
